### ISSUE TITLE
NPM Audit issue - Arbitrary File Overwrite

### ISSUE THREAD

**Issue by ma-jahn** [nodejs/node-gyp#1717]
* **Node Version**: node@v10.7.0 && npm@6.4.1
* **Platform**: OSX

<details></details>

NPM Audit issue:
Version 3.8.0 of node-gyp relies on tar < 4.4.2 which comes with a high Vulnerability (https://www.npmjs.com/advisories/803)

I can see that you have updated this dependency within your master branch, but version 4.0.0 is not yet released.

Can you let me know if and when this fix will be available?

Thanks

**Comment by antoniopaisfernandes** [comment]
https://github.com/nodejs/node-gyp/pull/1713

**Comment by cberg-zalando** [comment]
Following the comment stream in #1713 does not clearly say whether the fix will be backported into the v3.x branch. @richardlau just mentioned that the update of tar from version `2.*` to `3.*` was considered a breaking change at that time so it was only applied to master.

**Comment by StickieBE** [comment]
For those that wish:
mohsenari
https://github.com/sass/node-sass/issues/2625#issuecomment-482224111
&
osushi-desushi
https://github.com/sass/node-sass/issues/2625#issuecomment-482519446

Both posted a temporary solution to upgrade tar to ease the pain while we patiently wait for the update.

KR

**Comment by romanstingler** [comment]
It's a shame that this is known since april 4th and there is still no release.
Furthermore, it should be fixed a year ago when the issue was found in node-tar April 30th 2018.

You guys with your infinite number of dependencies should try a rolling release, as soon as one dependency is updatable you should upgrade and don't stick with node v4 which hasn't been updated since march 2018 so security updates or whatsoever.

In my opinion merge #1670 and #1718 and release this piece of .. 
and don't waste time in discussions if there should be a major or minor version bump, as it breaks compatibility with node 4 there must be a major release bump.

see semver.org

**Comment by flyke** [comment]
This is how I fixed this in my project, please let me know if there is a better way:
`npm install -D node-gyp`
`npm install -D tar@">4.4.7"`

Edit package-lock.json and replace:
`
"tar": {
          "version": "2.2.1",
          "resolved": "https://registry.npmjs.org/tar/-/tar-2.2.1.tgz",
          "integrity": "sha1-jk0qJWwOIYXGsYrWlK7JaLg8sdE=",
          "dev": true,
          "requires": {
            "block-stream": "*",
            "fstream": "^1.0.2",
            "inherits": "2"
          }
        }
`

with:

`
"tar": {
          "version": "4.4.8",
          "resolved": "https://registry.npmjs.org/tar/-/tar-4.4.8.tgz",
          "integrity": "sha512-LzHF64s5chPQQS0IYBn9IN5h3i98c12bo4NCO7e0sGM2llXQ3p2FGC5sdENN4cTW48O915Sh+x+EXx7XW96xYQ==",
          "dev": true,
          "requires": {
            "chownr": "^1.1.1",
            "fs-minipass": "^1.2.5",
            "minipass": "^2.3.4",
            "minizlib": "^1.1.1",
            "mkdirp": "^0.5.0",
            "safe-buffer": "^5.1.2",
            "yallist": "^3.0.2"
          }
        }
`

**Comment by ma-jahn** [comment]
@flyke Have a look at: https://github.com/nodejs/node-gyp/pull/1713
Updating tar comes with a breaking change, therefor manually overwriting the tar version in your package-lock.json will most likely break node-gyp and should therefore imho not be recommended solution.

**Comment by flyke** [comment]
@ma-jahn I am very open to any suggestion that fixes major security flaws like this. So if you have any 'good way' to fix this, please share, I would be very grateful.

**Comment by ma-jahn** [comment]
A fix is currently in progress: https://github.com/nodejs/node-gyp/commit/1456ef27c72cc4e49f9b1d82cbb047cb673156c8

**Comment by I-keep-trying** [comment]
> A fix is currently in progress: 1456ef2

Is there any reason a person couldn't, in the meantime, when installing node-sass, change package.json to include node-gyp 4.0.0, which already includes the updated tar version?

**Comment by JoshRobertson** [comment]
> 1456ef2

That fix is for 4.0.0 which hasn't been released to NPM yet.

This is the PR to release a 3.x update :
https://github.com/nodejs/node-gyp/pull/1718

This is the ticket regarding 4.0.0 not published to npm:
https://github.com/nodejs/node-gyp/issues/1721

**Comment by I-keep-trying** [comment]
So, nobody knows what we are supposed to do in the meantime? Just don't use node-sass? I'm not trying to be hyperbolic, I honestly don't know how serious this is.

**Comment by sidhujaspreet** [comment]
Just switched to node-sass for current project and ran into the same issue. Seems like it's been there since few days now, please pick this one up on priority.
Would be good to have rolling release as suggested by @romanstingler .

**Comment by jhnferraris** [comment]
It's been 8 days and not a single maintainer noticed this? Did you lose your npmjs account or something?

**Comment by corbin88** [comment]
>That fix is for 4.0.0 which hasn't been released to NPM yet.

Anyone know when this will be?

**Comment by kaiyoma** [comment]
Looks like a fix was merged 10 days ago, but there's still no release. What's the holdup?

**Comment by uniibu** [comment]
Temporary Solution for users using **Yarn**: (use at your own risk)

This is just a temporary solution for users using **Yarn** until the the fix gets published to npmjs.
- Verify that you affected by the vulnerability `yarn audit`
- If you are, edit package.json and add (do not install tar yet)
```
  "resolutions": {
    "tar": "^4.4.8"
  }
```
- After adding resolutions to your `package.json`, Install the latest tar package `yarn add tar@^4.4.8`
- Test it `yarn audit`  should no longer show that node-gyp -> tar vulnerability.

Once the fix has been published, simply remove both the `resolutions` and the package `yarn remove tar`  in the same order to update the `yarn.lock`

For `npm` users, it require manually editing you `package-lock.json`  and replacing the tar dependency with the latest. I don't recommend doing this if you don't know what you are doing. There are packages that could help you with this such as `https://www.npmjs.com/package/npm-force-resolutions`. Again use at your own risk.

**Comment by shazilrehman** [comment]
**npm install npm@latest -g**

that worked for me

**Comment by jhnferraris** [comment]
@shazilrehman What should be the correct npm version to be use? I am at 6.9.0 and still experiencing the problem. :D

**Comment by subhashkonda** [comment]
Do we have the fix ready for this, "npm audit fix" or explicit "npm install tar" is not helping in this case.

**Comment by juliusl** [comment]
@uniibu To clarify the risk, this will break older versions of node. Besides that this is basically what the current fix we're waiting on does.

I don't think you need the `yarn add tar@^4.4.8` though. Yarn seems to understand the resolutions on it's own and seems to update the yarn.lock appropriately.

**Comment by shazilrehman** [comment]
@jhnferraris may god be with you then.

**Comment by romanstingler** [comment]
It's a shame that you call yourself "Maintainer"
It's nice to see that yarn includes an easy way to get rid of this.
It's even better so see that other projects are moving away from your stuff 
[sweetalert2](https://github.com/sweetalert2/sweetalert2/commit/3fd16c42613ff2035bef119a18ffe08c7b97c3f2)

**Comment by benjamingr** [comment]
@romanstingler this sort of conduct is not appropriate in the Node.js org. Writing:

> It's a shame that you call yourself "Maintainer"

Is completely inappropriate here and not a legitimate way to vent your anger.

Please edit your comment and [conduct yourself according to how we expect users to interact here](https://github.com/nodejs/admin/blob/master/CODE_OF_CONDUCT.md).

**Comment by romanstingler** [comment]
Here is the security disclosure policy for Node.js

On the embargo date, the Node.js security mailing list is sent a copy of the announcement. **The changes are pushed to the public repository and new builds are deployed to nodejs.org**. Within 6 hours of the mailing list being notified, a copy of the advisory will be published on the Node.js blog.

This process can take some time, especially when coordination is required with maintainers of other projects. **Every effort will be made to handle the bug in as timely a manner as possible**; however, it’s important that we follow the release process above to ensure that the disclosure is handled in a consistent manner.


NOT HAPPEND !!!

**Comment by benjamingr** [comment]
A kind reminder that Node.js has interaction standards and maintainer abuse will not be tolerated. Comments that are not constructive will be deleted (criticism of technical issues and discussion is welcome).

If you are unsure how to interact we have a code of conduct [here](https://github.com/nodejs/admin/blob/master/CODE_OF_CONDUCT.md). If you have any questions about our policy or would like to report anything feel free to reach out at report@nodejs.org or moderation@nodejs.org

**Comment by ChALkeR** [comment]
This should better be fixed on node-tar side: https://github.com/npm/node-tar/issues/212

The ideal solution is node-tar releasing the fix as a semver-patch (that's possible) on all actively used branches of node-tar. That way, everyone would get the update seamlessly without any extra action needed.

Updating node-tar version in node-gyp would be a semver-major for node-gyp because of node-tar supported platforms change, and everyone would have the exact same problems updating `node-gyp`, then updating whatever depends on node-gyp (probably also a semver-major), and so on. That approach is problematic and will cause major ecosystem churn.

Also, this vuln is actually a false positive for node-gyp afaik, it should not be possible to exploit node-gyp with this in any other way that was not possible without this vuln.

**Comment by ChALkeR** [comment]
I think this is now fixed per https://github.com/nodejs/node-gyp/pull/1718#issuecomment-486023253 (thanks, @rvagg!).

Closing this now, feel free to reopen and/or comment though.

**Comment by samiragreco** [comment]
Hi!
I've update node.js and node-gyp but I still have the same issue.

```
 + node-gyp@4.0.0
added 9 packages from 2 contributors and audited 3975 packages in 3.189s
found 1 high severity vulnerability
run `npm audit fix` to fix them, or `npm audit` for details
```

**Comment by ChALkeR** [comment]
@samiragreco What does `npm audit` show? Could you do a gist?

**Comment by samiragreco** [comment]
@ChALkeR Thank you very much to answer! I'm a very new to programming, I hope this gist is what you meant. 

https://gist.github.com/samiragreco/6b61106f2077eb206f3ea635359fc0b1

**Comment by ma-jahn** [comment]
@samiragreco
In your case Node-gyp is actually a subdependency of node-sass. So the issue seems to be resolved for Node-gyp with the release of v4, but node-sass has to update its dependencies as well. Unfortunately this looks like a breaking change for them, so don't expect a update very soon... :(

**Comment by samiragreco** [comment]
Oh, ok . Thank you @ma-jahn , I see... What would happen if I continue a project with this vulnerability? (just a follow-along project for study). I can publish it on github? (Sorry just ignorance about this subject).

**Comment by benjamingr** [comment]
@samiragreco hey, in this particular case you can follow the advisory link (to https://nodesecurity.io/advisories/803  ) - this explains what the vulnerability is:

> Versions of tar prior to 4.4.2 are vulnerable to Arbitrary File Overwrite. Extracting tarballs containing a hardlink to a file that already exists in the system, and a file that matches the hardlink will overwrite the system's file with the contents of the extracted file.

This means that for example running a server that uses `tar` and accepts user input is vulnerable. In your specific case you should be fine (though update as soon as possible).

Asking this is entirely appropriate and you are welcome to raise or discuss these issues at https://github.com/nodejs/security-wg/issues

**Comment by samiragreco** [comment]
Thank you so much @benjamingr ! I really appreciate your giving your time to answer me! You have no idea how much you have helped me! :D

**Comment by JamieCorkhill** [comment]
@benjamingr @ChALkeR @ma-jahn 

Updating `node-gyp` does not fix the issue for me either. I've been having this problem ever since updating the `node-gyp configure.js` file as follows due to an issue with MSBuild.exe after updating to Visual Studio 2019 (lines 152-162):

```javascript
if (vsSetup) {
      // GYP doesn't (yet) have support for VS2017, so we force it to VS2015
      // to avoid pulling a floating patch that has not landed upstream.
      // Ref: https://chromium-review.googlesource.com/#/c/433540/
      gyp.opts.msvs_version = '2015'
      process.env['GYP_MSVS_VERSION'] = 2015
      process.env['GYP_MSVS_OVERRIDE_PATH'] = vsSetup.path
      defaults['msbuild_toolset'] = 'v142'
      defaults['msvs_windows_target_platform_version'] = vsSetup.sdk
      variables['msbuild_path'] = path.join(vsSetup.path, 'MSBuild', 'Current', 'Bin', 'MSBuild.exe')
    }
```
This may just be coincidental because I would think those changes would be irrelevant to `tar`.

Here is my NPM Audit Report:

https://gist.github.com/JamieCorkhill/90f92c4f50d7e59e42c03a90d11ad93c

Any help would be much appreciated.

Thanks.

**Comment by subhashkonda** [comment]
Any update on this fix, I saw some where the ETA was 10th May, so thought check again on new ETA ???

**Comment by ChALkeR** [comment]
@JamieCorkhill 

As your report shows, `npm` in your setup chain-depends on old node-gyp version, which depends on old `tar` version. For this to be fixed, either of those two should happen: 
* your dependencies need to migrate to a new node-gyp version (which is on npm, Inc. side), or
* `tar` should release a semver-patch version with the fix at 2.x branch that would be compatible to the version contract of the old `node-gyp`.

That is not a problem of node-gyp and isn't something that is trivially fixable on node-gyp side.

**Comment by JamieCorkhill** [comment]
@ChALkeR 

Thank you. So, to confirm, `npm` itself in my dependency chain relies on an old version of `node-gyp` which relies on an old version of `tar`? I updated Node and `npm` yesterday to their newest versions, and have had the issue since then.

**Comment by ChALkeR** [comment]
@JamieCorkhill Yes, you can confirm that it does not happen with `npm install node-gyp` in a clean project directory but happens with `npm install npm`.

**Comment by JamieCorkhill** [comment]
@ChALkeR Thanks. So are `npm` aware of this issue do you know of, and have they taken any steps to mitigate it?

**Comment by ChALkeR** [comment]
@JamieCorkhill There are four separate dep chains in `npm` leading to the outdated `tar` dependency in the audit you linked above. It appears that those are being worked on, e.g. `npm-lifecycle` had a release recently (5 days ago) that bumped its deps: https://github.com/npm/npm-lifecycle/commits/latest, and now at least `npm lifecycle` does not cause audit warnings anymore.

It is understandable why unrolling those deps chains is not fast, if they want to be on the safe side and test everything.

That said, I would have appreciated slightly more feedback in the tar@2 backport issue/pr :-).

**Comment by ChALkeR** [comment]
I think this is finally fixed with everyone with `node-tar@2.2.2` release, except for packages where exact dependency versions are locked.
